server/lib/parties/party-service.ts

```diff
@@ -0,0 +1,228 @@
+  members: Map<number, PartyUser>
+  leaderId: number
+}
+
+export enum PartyServiceErrorCode {
+  PartyNotFound,
+  InsufficientPermissions,
+  PartyFull,
+  UserOffline,
+}
+
+export class PartyServiceError extends Error {
+  public code: PartyServiceErrorCode
+
+  constructor(code: PartyServiceErrorCode, message: string) {
+    super(message)
+    this.code = code
+  }
+}
+
+export function getInvitesPath(partyId: string): string {
+  return `/parties/invites/${partyId}`
+}
+
+export function getPartyPath(partyId: string): string {
+  return `/parties/${partyId}`
+}
+
+@singleton()
+export default class PartyService {
+  private parties = new Map<string, PartyRecord>()
+  private clientSocketsToPartyId = new Map<ClientSocketsGroup, string>()
+
+  constructor(
+    private nydus: NydusServer,
+    private clientSocketsManager: ClientSocketsManager,
+    private userSocketsManager: UserSocketsManager,
+  ) {}
+
+  invite(leader: PartyUser, leaderClientId: string, invites: PartyUser[]): Readonly<PartyRecord> {
+    const leaderClientSockets = this.getClientSockets(leader.id, leaderClientId)
+
+    let party = this.getClientParty(leaderClientSockets)
+    if (party) {
+      if (party.leaderId !== leader.id) {
+        throw new PartyServiceError(
+          PartyServiceErrorCode.InsufficientPermissions,
+          'Only party leader can invite people',
+        )
+      }
+
+      for (const invite of invites) {
+        party.invites.set(invite.id, invite)
+      }
+
+      this.publishToParty(party.id, {
+        type: 'invite',
+        invites,
+      })
+    } else {
+      const partyId = cuid()
+      party = {
+        id: partyId,
+        invites: new Map(invites.map(i => [i.id, i])),
+        members: new Map([[leader.id, leader]]),
+        leaderId: leader.id,
+      }
+
+      this.parties.set(partyId, party)
+      this.clientSocketsToPartyId.set(leaderClientSockets, partyId)
+      this.subscribeToParty(leaderClientSockets, party)
+    }
+
+    invites
+      .map(i => this.getUserSockets(i.name))
+      .forEach(userSockets => {
+        userSockets.subscribe(
+          getInvitesPath(party!.id),
+          () => ({
+            type: 'invite',
+            from: party!.leaderId,
+          }),
+          () => {
+            // TODO(2Pac): Handle user quitting; need to keep a map of user -> invites?
+          },
+        )
+      })
+
+    return party
+  }
+
+  removeInvite(partyId: string, target: PartyUser, leader?: PartyUser) {
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound, 'Party not found')
+    }
+
+    if (leader && leader.id !== party.leaderId) {
+      throw new PartyServiceError(
+        PartyServiceErrorCode.InsufficientPermissions,
+        'Only party leaders can remove invites to other people',
+      )
+    }
+
+    // TODO(2Pac): Check if the target is invited first before trying to do any of the below?
```

Comment:
`delete` returns true/false to say whether anything was actually removed, so you can just use that + throw if it's false?